### PR TITLE
For #17: created test for filelist creation option

### PR DESCRIPTION
**Pull request by paulodamaso**
For #17: created test for filelist creation option
- Created field and changed constructors in `Rpm` to accept filelist creation flag
- Added test which checks if `filelist.xml` and `filelist.xml.gz` files are created
- Left puzzle to implement the feature

### CODE REVIEW

**Comment by 0crat** [comment]
This pull request [#42](https://github.com/artipie/rpm-adapter/issues/42) is assigned to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/3c70720c-8320-49c1-ba1e-534ed5cd9242); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

**Comment by paulodamaso** [comment]
@amihaiemil PLease review, seems that the failures aren't directly related to my changes

**Comment by 0crat** [comment]
@amihaiemil[/z](https://www.0crat.com/u/amihaiemil) this job was assigned to you 5days ago. It will be taken away from you soon, unless you close it, see [§8](http://www.zerocracy.com/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

**Comment by amihaiemil** [comment]
@0crat refuse

**Comment by 0crat** [comment]
> @0crat refuse ([here](https://github.com/artipie/rpm-adapter/issues/42#issuecomment-602587065))

@amihaiemil The user @amihaiemil[/z](https://www.0crat.com/u/amihaiemil) resigned from [#42](https://github.com/artipie/rpm-adapter/issues/42), please stop working. Reason for job resignation: Order was cancelled

**Comment by 0crat** [comment]
This pull request [#42](https://github.com/artipie/rpm-adapter/issues/42) is assigned to @vatavuk[/z](https://www.0crat.com/u/vatavuk), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/82f37d82-e02e-436d-af2e-29f4733d5ef1); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

**Comment by Vatavuk** [comment]
@paulodamaso you have conflicts

File: src/test/java/com/artipie/rpm/RpmTest.java [hunk 1]
```diff
@@ -63,27 +67,62 @@
...
+        @TempDir final Path folder, @TempDir final Path store
+    ) throws Exception {
+        final String key = "no-fileList.aarch64.rpm";
+        final Storage storage = RpmTest.save(folder, store, key);
+        new Rpm(storage).update(key).blockingAwait();
+        MatcherAssert.assertThat(
+            storage.list(new Key.From(RpmTest.REPODATA))
+                .get().stream().map(Key::string).collect(Collectors.toList()),
+            new IsIterableContainingInAnyOrder<>(
+                new ListOf<Matcher<? super String>>(
```
Comment (Vatavuk) [inline src/test/java/com/artipie/rpm/RpmTest.java]: @paulodamaso we should assert that these files do not exist.

File: src/main/java/com/artipie/rpm/Rpm.java [hunk 2]
```diff
@@ -81,25 +87,32 @@
...
     }
 
     /**
      * Ctor.
      * @param stg The storage
      * @param naming RPM files naming policy
+     * @param filelistsgen Flag to generate the filelist.
      */
-    public Rpm(final Storage stg, final NamingPolicy naming) {
+    public Rpm(final Storage stg, final NamingPolicy naming, final boolean filelistsgen) {
```
Comment (Vatavuk) [inline src/main/java/com/artipie/rpm/Rpm.java]: @paulodamaso Do we have an alternative to a boolean parameter? Using this class is not user friendly:
` new Rpm(storage, NamingPolicy.DEFAULT, false)` How about using `factory` classes or wrapping the boolean to an object?

**Comment by paulodamaso** [comment]
@Vatavuk Thanks for the review. I've fixed the boolean flag problem throgh the use of decoration, please take a look. As for the test, due changes in introduced in #46 `Rpm` does not create the files anymore, so the test now is correct. I just moved it to the correct class, and it is still disabled because the behavior is not implemented yet.

**Comment by paulodamaso** [comment]
@Vatavuk Ping

File: src/main/java/com/artipie/rpm/WithFilelists.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+ * SOFTWARE.
+ */
+package com.artipie.rpm;
+
+import io.reactivex.Completable;
+
+/**
+ * Rpm decorator which provide filelist creation on update.
+ *
+ * @since 0.1
```
Comment (Vatavuk) [inline src/main/java/com/artipie/rpm/WithFilelists.java]: @paulodamaso since 0.3.3

File: src/test/java/com/artipie/rpm/WithFilelistsTest.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+import org.hamcrest.collection.IsIterableContainingInAnyOrder;
+import org.hamcrest.core.IsEqual;
+import org.junit.jupiter.api.Disabled;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.io.TempDir;
+
+/**
+ * Test for {@link WithFilelists}.
+ *
+ * @since 0.1
```
Comment (Vatavuk) [inline src/test/java/com/artipie/rpm/WithFilelistsTest.java]: @paulodamaso since 0.3.3

**Comment by paulodamaso** [comment]
@Vatavuk Done, please take a look

**Comment by g4s8** [comment]
@paulodamaso please resolve merge conflicts

**Comment by paulodamaso** [comment]
@g4s8 Done!

**Comment by Vatavuk** [comment]
@g4s8 ping

**Comment by Vatavuk** [comment]
@paulodamaso confilcts

**Comment by paulodamaso** [comment]
@g4s8 done

**Comment by 0crat** [comment]
@vatavuk[/z](https://www.0crat.com/u/vatavuk) this job was assigned to you 5days ago. It will be taken away from you soon, unless you close it, see [§8](http://www.zerocracy.com/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

**Comment by paulodamaso** [comment]
@Vatavuk Ping

**Comment by Vatavuk** [comment]
@g4s8 we can merge

**Comment by Vatavuk** [comment]
@g4s8 pls merge or refuse, I have only two days left to get paid.

**Comment by paulodamaso** [comment]
@g4s8 please merge

**Comment by g4s8** [comment]
@Vatavuk 0crat won't resign you on PRs, 10 days rule works only for issues, not PRs

**Comment by paulodamaso** [comment]
> @paulodamaso there are some concerns about this PR: RPM adapter is already in use by our clients, so we can't change its API completely, it leads to big refactoring on their side. Is it possible to hide this implementation under some additional constructor params, e.g. `new Rpm(storage, new FilelistsOptions(/*include: */: true))` or something similar and keep current secondary constructors as is without changes?

@g4s8 It was my first implementation, but me and @Vatavuk agreed that it wasn't correctly designed. What if I:
- implement this design I'm proposing with another naming
- wrap it with current `Rpm` to avoid client code refactor
- add a constructor to `Rpm` which would be used to create adapter with filelist option
- fix tests (obviously)

I think that this way we could at least keep a clean design in the API core and avoid client code refactoring, WDYT?

**Comment by g4s8** [comment]
> > @paulodamaso there are some concerns about this PR: RPM adapter is already in use by our clients, so we can't change its API completely, it leads to big refactoring on their side. Is it possible to hide this implementation under some additional constructor params, e.g. `new Rpm(storage, new FilelistsOptions(/*include: */: true))` or something similar and keep current secondary constructors as is without changes?
> 
> @g4s8 It was my first implementation, but me and @Vatavuk agreed that it wasn't correctly designed. What if I:
> 
>     * implement this design I'm proposing with another naming
> 
>     * wrap it with current `Rpm` to avoid client code refactor
> 
>     * add a constructor to `Rpm` which would be used to create adapter with filelist option
> 
>     * fix tests (obviously)
> 
> 
> I think that this way we could at least keep a clean design in the API core and avoid client code refactoring, WDYT?

@paulodamaso yes, it should work

**Comment by paulodamaso** [comment]
@g4s8 Could you please boost it? I've remade this PR several times due to conflict and will have to make it again due this architectural change

**Comment by paulodamaso** [comment]
@g4s8 Done, please take a look

**Comment by g4s8** [comment]
@paulodamaso please check build failures

**Comment by paulodamaso** [comment]
@g4s8 Done, please take a look

**Comment by 0crat** [comment]
@sereshqua[/z](https://www.0crat.com/u/sereshqua) please review this job completed by @vatavuk[/z](https://www.0crat.com/u/vatavuk), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
Code review was too long (11 days), architects (@g4s8) were penalized, see [§55](http://www.zerocracy.com/policy.html#55)

**Comment by sereshqua** [comment]
@0crat quality good

**Comment by Sammers21** [comment]
@paulodamaso , 480 lines of the diff...

**Pull request merged**